[VideoPlayer] Close PGS/VobSub subtitles when audio language dont match

CastagnaIT opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
Stream selection always try to select the "best" track based on settings also when subtitles are disabled, and internally the stream is kept opened

This means that even though subtitles are disabled, they are still parsed.
Parsing PGS/VobSub subtitles that support “forced” mode on individual images will override the subtitle settings,
with the result that "forced" images are always displayed on screen, even if subtitles settings are set to disabled.

This can also lead to the display of an unwanted language,
the behaviour is:
`-> kodi select "best" language -> open the stream -> if dont match the settings, it set the subs visibility to false, but still kept the stream opened`

so it is necessary to close the stream at least when it does not match the audio language,
in order to avoid parsing (and show on screen) unwanted languages

this is based on closed PR #26087 after further verification i think this is currently the best approach
except that i added an additional check for the subtitle type so as not to affect other subtitle types with this behavior

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
fix #26050
## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
user test stream:
[Bud_VOBSUB_Forced.zip](https://github.com/user-attachments/files/23499113/Bud_VOBSUB_Forced.zip)

1) set kodi audio and subtitles language settings to english
2) play the video
3) BEFORE: kodi show unwanted language (forced overlays)
AFTER: kodi internally close the stream because dont match audio language (forced overlays are not parsed/shown)

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
avoid show unwanted language
## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

thexai reviewed (approved):
👍

garbear commented:
@claude please review

CastagnaIT commented:
i done some more tests (always by deleting MyVideos*.db database between tests to avoid have settings stored that invalidate tests)
1) (sub lang set to english) -> play sample -> subs are disabled -> OSD enable subs -> subs are displayed correctly
2) (sub lang set to english) -> play sample -> subs are disabled -> OSD enable and disable subs -> subs are not displayed correctly
3) (sub lang set to english) -> play sample -> subs are disabled -> OSD enable and disable and re-enable subs -> subs are displayed correctly

this change seem dont cause side effects, or please specify better the use case, maybe i didn't understand...

However, if you are considering to fully disabling “forced” overlays, this is not covered by this PR
f.e. VLC dont support "forced overlays" just it's just a simple enable/disable subs (at least it seems that way)

as @thexai said PGS/Vobsub should always have the "forced" subs displayed regardless of the settings (this is by subtitles type specs)
if instead we want provide a way to disable "forced" overlay support, i agree that another setting is required

garbear commented on xbmc/utils/StreamUtils.cpp:
```diff
@@ -150,3 +150,9 @@ std::string StreamUtils::GetLayout(unsigned int channels)
+bool StreamUtils::IsCodecSupportForcedOverlay(int codecId)
+{
+  return codecId == AV_CODEC_ID_DVD_SUBTITLE || codecId == AV_CODEC_ID_HDMV_PGS_SUBTITLE ||
+         codecId == AV_CODEC_ID_DVB_SUBTITLE;
```
`AV_CODEC_ID_DVB_SUBTITLE` is listed twice?

garbear replied:
You're right. I even tried comparing pixels and missed the difference. Not sure if code can be made more clear but it's correct as is.

garbear commented:
@claude checking in, where are we at?

CastagnaIT commented:
> Yes the settings are a bit different, I tested with default settings. With a new portable directory, they are:
> <img alt="image" width="1226" height="236" src="https://private-user-images.githubusercontent.com/489377/517581253-897acdf4-6349-4f2a-8916-450d379fc57d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM4ODkxNzAsIm5iZiI6MTc2Mzg4ODg3MCwicGF0aCI6Ii80ODkzNzcvNTE3NTgxMjUzLTg5N2FjZGY0LTYzNDktNGYyYS04OTE2LTQ1MGQzNzlmYzU3ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMTIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTEyM1QwOTA3NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03NTkxOTQyOGI0OTY5MzU1ZTVjMmI2MGEwMTc0MTY1Y2RkNzNlNjBlYTUxZjZmNjQ5MmNkNzM3NWJlMTE5YmZhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.9bEZCVqN5Z8QUpXzt--XafMTkVLd2f1WTqYpNSfpfA0"> <img alt="image" width="1368" height="345" src="https://private-user-images.githubusercontent.com/489377/517564302-5508b39d-c48a-465f-81f3-ba2ae2e911cc.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM4ODkxNzAsIm5iZiI6MTc2Mzg4ODg3MCwicGF0aCI6Ii80ODkzNzcvNTE3NTY0MzAyLTU1MDhiMzlkLWM0OGEtNDY1Zi04MWYzLWJhMmFlMmU5MTFjYy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMTIzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTEyM1QwOTA3NTBaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jMjM4NzUyZDY0YmM1NGYyOWFhMGY4YmIxNDAyOTQ3MmU1ZGFhZjJjNGZmODNjM2I0ODFhYWZiMWM2YWVmOTZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.O4W3ywhnhAyTvr5LI5Xq5mF_h5t76Q4jtB_d8zldqnY">
> 
> Play the sample: subs not displayed. Toggle subs on / press T (and seek back for the short sample): toast popup with "Subtitles Swedish", but no subs rendered. After selecting a different subs track (and seeking back), they appear (even the Swedish that were initially not displayed)
> 
> The default preferred sub language was changed recently; reverting to "none" gives the same result.

ok i reproduced it
so by using GUI subtitles are shown correctly
but if you use keyboard (with T) subtitles are NOT shown correctly
as soon as i have time, i will try to see if i can understand it better this difference

CastagnaIT committed:
[VideoPlayer] Close PGS/VobSub subtitles when language dont match

Stream selection always try to select the "best" track based on settings
also when subtitles are disabled, and internally the stream is kept opened.
This means that even though subtitles are disabled, they are still parsed. Parsing PGS/VobSub subtitles that support “forced” mode on individual images will override the subtitle settings, with the result that they are displayed on screen even if disabled. This can also lead to the display of an unwanted language, so it is necessary to close the stream at least when it does not match the audio, in order to avoid parsing unwanted languages.
